### PR TITLE
Foorm: add pre/post/daily routes

### PR DESCRIPTION
**Pull request by molly-moen**
Add routes for summer workshop surveys. The routes are:
```
/pd/workshop_pre_survey
/pd/workshop_daily_survey/day/:day
/pd/workshop_post_survey
```
All routes can have `?enrollmentCode=CODE` at the end to specify a specific workshop enrollment, otherwise they will use the most recent workshop the user is enrolled in.

Also includes some refactoring of common code and an additional check when storing facilitator survey results (if the survey results for a facilitator have no answers, don't store the data).

### Future work
Next steps are ensuring our results views show the pre/daily/post surveys correctly.

## Links

- [jira](https://codedotorg.atlassian.net/browse/PLC-885)

## Testing story
Updated and added unit tests for workshop_daily_survey_controller. Also manually tested the new routes work and the csf intro post route still works, as it was refactored slightly.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/lib/pd/workshop_survey_foorm_constants.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+module Pd
+  module WorkshopSurveyFoormConstants
+    include SharedWorkshopConstants
```
Comment (bencodeorg) [inline dashboard/lib/pd/workshop_survey_foorm_constants.rb:3]: Do these get used at all?

File: dashboard/app/controllers/pd/workshop_daily_survey_controller.rb [hunk 2]
```diff
@@ -298,21 +302,15 @@ def new_csf_post101
...
       return render :no_attendance unless attended_workshop
 
-      # Render a thanks message if already submitted.
-      if !params[:force_show] && Pd::WorkshopSurveyFoormSubmission.has_submitted_form?(current_user.id, attended_workshop.id, nil, nil, survey_name)
-        render :thanks
-        return
-      end
-
-      render_csf_survey_foorm(survey_name, attended_workshop)
+      render_survey_foorm(survey_name, attended_workshop, nil, nil)
```
Comment (breville) [inline dashboard/app/controllers/pd/workshop_daily_survey_controller.rb]: Maybe a bit more readable with named parameters?

File: dashboard/app/controllers/api/v1/pd/workshop_survey_foorm_submissions_controller.rb [hunk 3]
```diff
@@ -1,11 +1,12 @@
 class Api::V1::Pd::WorkshopSurveyFoormSubmissionsController < ApplicationController
+  include Pd::WorkshopSurveyFoormConstants
   def create
     answers = params[:answers]
 
     # save facilitator answers as separate survey submissions
     # for ease of querying per-facilitator data
-    facilitator_answers = answers["facilitators"]
-    answers.delete("facilitators")
+    facilitator_answers = answers[FACILITATORS]
```
Comment (breville) [inline dashboard/app/controllers/api/v1/pd/workshop_survey_foorm_submissions_controller.rb]: Would it be slightly clearer where `FACILITATORS` came from if it had a namespace or some other kind of prefix?

File: dashboard/lib/pd/workshop_survey_foorm_constants.rb (hunk 1 above)
Comment (molly-moen) [reply dashboard/lib/pd/workshop_survey_foorm_constants.rb:3]: nope, I had used them then changed it, removing

File: dashboard/app/controllers/api/v1/pd/workshop_survey_foorm_submissions_controller.rb (hunk 3 above)
Comment (molly-moen) [reply dashboard/app/controllers/api/v1/pd/workshop_survey_foorm_submissions_controller.rb]: I can prefix with Pd::WorkshopSurveyFoormConstants if that makes it clearer. Should there be a prefix on the variable name itself instead?

File: dashboard/lib/pd/workshop_survey_foorm_constants.rb (hunk 1 above)
Comment (molly-moen) [reply dashboard/lib/pd/workshop_survey_foorm_constants.rb:3]: just kidding, they are used. SUBJECT_SUMMER_WORKSHOP, etc are from SharedWorkshopConstants

File: dashboard/lib/pd/workshop_survey_foorm_constants.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    DAILY_SURVEY_NAMES = {
```
Comment (bencodeorg) [inline dashboard/lib/pd/workshop_survey_foorm_constants.rb]: Is "names" the right word for this? Like `DAILY_SURVEY_CONFIG_PATHS` or something?

Comment (molly-moen) [reply dashboard/lib/pd/workshop_survey_foorm_constants.rb]: good call

File: dashboard/app/controllers/pd/workshop_daily_survey_controller.rb (hunk 2 above)
Comment (molly-moen) [reply dashboard/app/controllers/pd/workshop_daily_survey_controller.rb]: done!

File: dashboard/app/controllers/api/v1/pd/workshop_survey_foorm_submissions_controller.rb [hunk 5]
```diff
@@ -52,13 +53,16 @@ def save_facilitator_questions(facilitator_answers, params)
     survey_submission_ids = []
     if facilitator_answers
       facilitator_answers.each do |_, data|
-        next unless data["facilitator_id"]
+        # data needs to have facilitator id
+        next unless data[FACILITATOR_ID]
+        # If data only contains metadata, do not store, as there were no answers for this facilitator.
+        next if data.except(FACILITATOR_ID, FACILITATOR_NAME, FACILITATOR_POSITION).nil_or_empty?
```
Comment (bencodeorg) [inline dashboard/app/controllers/api/v1/pd/workshop_survey_foorm_submissions_controller.rb]: Not necessary for this PR, but in the future maybe some way we can identify a group of metadata fields somewhere? I could imagine this list of metadata changing, then we start writing results when there isn't any real data.

File: dashboard/test/controllers/pd/workshop_daily_survey_controller_test.rb [hunk 6]
```diff
@@ -105,8 +105,8 @@ class WorkshopDailySurveyControllerTest < ActionDispatch::IntegrationTest
         user: @enrolled_summer_teacher
 
       sign_in @enrolled_summer_teacher
-      get '/pd/workshop_survey/foorm/day/0'
-      assert_redirected_to action: 'thanks'
+      get '/pd/workshop_pre_survey'
+      assert_thanks
     end
 
     test 'pre-workshop survey displays embedded JotForm when enrolled' do
```
Comment (bencodeorg) [inline dashboard/test/controllers/pd/workshop_daily_survey_controller_test.rb:112]: Rename?

**Review by bencodeorg — APPROVED** [review]
LGTM! One question for my understanding -- the logic that figures out what survey to show is so that we can render a) in cases when we have a specific enrollment code (eg, maybe in a workshop mailer?) and b) when a teacher is logged in, but visits a survey url directly (eg, maybe shared by a workshop organizer)?

File: dashboard/app/controllers/api/v1/pd/workshop_survey_foorm_submissions_controller.rb (hunk 3 above)
Comment (breville) [reply dashboard/app/controllers/api/v1/pd/workshop_survey_foorm_submissions_controller.rb]: I think the namespace might be helpful enough.

**Review by breville — APPROVED** [review]

**Comment by molly-moen** [comment]
> LGTM! One question for my understanding -- the logic that figures out what survey to show is so that we can render a) in cases when we have a specific enrollment code (eg, maybe in a workshop mailer?) and b) when a teacher is logged in, but visits a survey url directly (eg, maybe shared by a workshop organizer)?

@bencodeorg Yep that's correct!

File: dashboard/test/controllers/pd/workshop_daily_survey_controller_test.rb (hunk 6 above)
Comment (molly-moen) [reply dashboard/test/controllers/pd/workshop_daily_survey_controller_test.rb:112]: the Jotform test? This test still refers to the Jotform link

Comment (bencodeorg) [reply dashboard/test/controllers/pd/workshop_daily_survey_controller_test.rb:112]: Oops, github hidden unchanged lines confused me. Ignore!

**Pull request merged**